state-reconstruct-fetcher/src/l1_fetcher.rs

```diff
@@ -124,26 +124,56 @@ impl L1Fetcher {
         })
     }
 
-    pub async fn run(&self, sink: mpsc::Sender<CommitBlock>) -> Result<()> {
+    /// Decide which block to start fetching from based on the following criteria:
+    /// - Has the tool already made progress before?
+    /// - Was a snapshot just imported?
+    /// - Did the user set an explicit start block?
+    ///
+    /// Returns the block number to start fetching from.
+    async fn decide_start_block(&self, snapshot_end_batch: Option<U64>) -> Result<U64> {
         // Start fetching from the `GENESIS_BLOCK` unless the `start_block` argument is supplied,
         // in which case, start from that instead. If no argument was supplied and a state snapshot
         // exists, start from the block number specified in that snapshot.
-        let mut current_l1_block_number = U64::from(self.config.start_block);
+        let mut start_block = U64::from(self.config.start_block);
+
+        // We also have to check if a snapshot was recently imported. If so we
+        // should continue from the last imported batch.
+        if let Some(target_batch_number) = snapshot_end_batch {
+            tracing::info!(
+                "Trying to map snapshots latest L1 batch number, this might take a while..."
+            );
+            match self.map_l1_batch_to_l1_block(U256::from(target_batch_number.as_u64())).await {
+                Ok(block_number) => return Ok(block_number),
+                Err(e) => tracing::error!("Unable to find a corresponding L1 block number for the latest imported L1 batch: {e}"),
```

Comment:
Why not just return, when `map_l1_batch_to_l1_block` returns error?

Reply:
In the case that the L1 block number can't be found, reconstruction will still work. It's just that the fetcher will start from the beginning and take longer to catch back up. So, since its recoverable from that state, I don't think we need to terminate the whole process.

Reply:
OK